### PR TITLE
[ZEPPELIN-1614] Remove markdown4j dep

### PR DESCRIPTION
**Pull request by 1ambda**
### What is this PR for?

Remove unmanaged, old library markdown4j dep which exists just for Websequence and YUML plugins. 
(related to https://github.com/apache/zeppelin/pull/1384)

By adding Websequence and YUML plugins to pegdown

- Removing markdown4j dependency which is unmanaged library currently. 
- Addtionally, we can remove `markdown.parser.type` options in **markdown interpreter**
- Fixing some bugs in Websequence and YUML plugins
- Enable others to add more plugins using pegdown sytnax.

### What type of PR is it?

Improvement

### Todos

Nothing

### What is the Jira issue?

[JIRA - ZEPPELIN-1614](https://issues.apache.org/jira/browse/ZEPPELIN-1614)

### How should this be tested?

Some functional tests are included.

### Questions:
* Does the licenses files need update? - YES
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - YES

### CODE REVIEW

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+public class PegdownWebSequencelPlugin extends Parser implements BlockPluginParser {
+
+  public PegdownWebSequencelPlugin() {
+    super(PegdownParser.OPTIONS,
+        PegdownParser.PARSING_TIMEOUT_AS_MILLIS,
+        DefaultParseRunnerProvider);
+  }
+
+  public PegdownWebSequencelPlugin(Integer options,
+                                   Long maxParsingTimeInMillis,
```
Comment (bzz) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: In general, Zeppelin project [Jave style guide](https://google.github.io/styleguide/javaguide.html) discourage using of [horizontal alignment](https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment)

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      String json = response.toString();
+
+      int start = json.indexOf("?png=");
+      int end = json.indexOf("\"", start);
+
+      if (start != -1 && end != -1) {
+        return "http://www.websequencediagrams.com/" + json.substring(start, end);
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get proper response from websequencediagrams.com");
```
Comment (bzz) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: It's always good idea to propagate e as cause though Constructor to aid debugging.

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      writer.close();
+      reader.close();
+
+      String json = response.toString();
+
+      int start = json.indexOf("?png=");
+      int end = json.indexOf("\"", start);
+
+      if (start != -1 && end != -1) {
+        return "http://www.websequencediagrams.com/" + json.substring(start, end);
```
Comment (bzz) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: Is there a reason why url string, hardcoded in multiple places, can not be replaced with the constant field?

**Comment by bzz** [comment]
Websequence are awesome, thank you @1ambda for contributing this simplification of MD interpreter.

Looks good to me, modulo few minor issues, commented above. 

As soon as CI is green and issues addressed - will be happy to merge to master, in case there is no further discussion.

**Comment by felixcheung** [comment]
This is awesome, thanks!
+1 on @bzz 's comments

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java (hunk 1 above)
Comment (1ambda) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: ✅

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java (hunk 2 above)
Comment (1ambda) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: ✅

File: markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java (hunk 3 above)
Comment (1ambda) [inline markdown/src/main/java/org/apache/zeppelin/markdown/PegdownWebSequencelPlugin.java]: ✅

File: docs/interpreter/markdown.md [hunk 4]
```diff
@@ -25,42 +25,25 @@ limitations under the License.
...
 <img src="../assets/themes/zeppelin/img/docs-img/markdown-example.png" width="70%" />
 
-### Markdown4j Parser
+The markdown interpreter provides github flavored markdown.
 
-`markdown4j` parser provides [YUML](http://yuml.me/) and [Websequence](https://www.websequencediagrams.com/) extensions 
+<img src="../assets/themes/zeppelin/img/docs-img/markdown-example-github-flavored-parser.png" width="70%" />
 
-<img src="../assets/themes/zeppelin/img/docs-img/markdown-example-markdown4j-parser.png" width="70%" />
+### Supproted Plugins
```
Comment (1ambda) [inline docs/interpreter/markdown.md]: :) ✅

**Comment by 1ambda** [comment]
Thanks for the review @bzz, @felixcheung! I'v just fixed :)

**Comment by Leemoonsoo** [comment]
Thanks @1ambda for the contribution.
I have compared output from markdown4j and pegdown and found some differences.

font difference
image

emphasis inside of `()`
image

There could be more unknown differences. That means, existing notebook might render differently with pegdown and that could be unexpected result in some cases.

So how about keep markdown4j and make pegdown default parser? That at least give user option to configure markdown interpreter with old markdown4j parser when they don't want to make change on existing notebook.

**Comment by Leemoonsoo** [comment]
And i think it'll be better mention in `docs/install/upgrade.html#upgrading-from-zeppelin-06-to-07` about default markdown parser change and result slightly different output rendering of markdown paragraphs. And link to markdown parser configuration section in the doc would help as well.

**Comment by 1ambda** [comment]
@Leemoonsoo Thanks for review 👍 

It would be better to

1. Use pegdown as default
2. Add docs for it
3. Revert markdown4j

Regarding to rendered output you provided, I think it is a bug of pegdown. 
But even if it is resolved, there might be differences we can miss. :( 
So keep markdown4j. (I overlooked backward compatibility issue)

**Comment by bzz** [comment]
What should be the next steps here @1ambda?

**Comment by 1ambda** [comment]
@bzz Resolving CI failure will be next step. But i think it is not problem of markdown processor. All tests are failed in scalding interpreter test

I will rebase and retrigger CI again.

```
// PROFILE="-Pspark-2.0 -Phadoop-2.3 -Ppyspark -Psparkr -Pscalding -Pexamples -Pscala-2.11"
// https://s3.amazonaws.com/archive.travis-ci.org/jobs/173808635/log.txt
// https://s3.amazonaws.com/archive.travis-ci.org/jobs/173808636/log.txt
// https://s3.amazonaws.com/archive.travis-ci.org/jobs/173808637/log.txt

[INFO] Reactor Summary:
[INFO] 
[INFO] Zeppelin ........................................... SUCCESS [ 12.258 s]
[INFO] Zeppelin: Interpreter .............................. SUCCESS [ 17.129 s]
[INFO] Zeppelin: Zengine .................................. SUCCESS [ 11.340 s]
[INFO] Zeppelin: Display system apis ...................... SUCCESS [ 19.516 s]
[INFO] Zeppelin: Spark dependencies ....................... SUCCESS [ 44.808 s]
[INFO] Zeppelin: Spark .................................... SUCCESS [ 23.799 s]
[INFO] Zeppelin: Markdown interpreter ..................... SUCCESS [  0.594 s]
[INFO] Zeppelin: Angular interpreter ...................... SUCCESS [  0.270 s]
[INFO] Zeppelin: Shell interpreter ........................ SUCCESS [  0.335 s]
[INFO] Zeppelin: Livy interpreter ......................... SUCCESS [02:34 min]
[INFO] Zeppelin: HBase interpreter ........................ SUCCESS [ 10.105 s]
[INFO] Zeppelin: Apache Pig Interpreter ................... SUCCESS [  5.611 s]
[INFO] Zeppelin: PostgreSQL interpreter ................... SUCCESS [  0.615 s]
[INFO] Zeppelin: JDBC interpreter ......................... SUCCESS [  1.026 s]
[INFO] Zeppelin: File System Interpreters ................. SUCCESS [  1.140 s]
[INFO] Zeppelin: Flink .................................... SUCCESS [  8.944 s]
[INFO] Zeppelin: Apache Ignite interpreter ................ SUCCESS [  1.168 s]
[INFO] Zeppelin: Kylin interpreter ........................ SUCCESS [  0.407 s]
[INFO] Zeppelin: Python interpreter ....................... SUCCESS [  0.397 s]
[INFO] Zeppelin: Lens interpreter ......................... SUCCESS [  4.180 s]
[INFO] Zeppelin: Apache Cassandra interpreter ............. SUCCESS [01:04 min]
[INFO] Zeppelin: Elasticsearch interpreter ................ SUCCESS [  8.569 s]
[INFO] Zeppelin: BigQuery interpreter ..................... SUCCESS [  4.676 s]
[INFO] Zeppelin: Alluxio interpreter ...................... SUCCESS [  3.305 s]
[INFO] Zeppelin: web Application .......................... SUCCESS [01:37 min]
[INFO] Zeppelin: Server ................................... SUCCESS [ 22.082 s]
[INFO] Zeppelin: Packaging distribution ................... SUCCESS [01:04 min]
[INFO] Zeppelin: Scalding interpreter ..................... FAILURE [  1.212 s]
[INFO] Zeppelin: Examples ................................. SKIPPED
[INFO] Zeppelin: Example application - Clock .............. SKIPPED
[INFO] Zeppelin: Example application - Horizontal Bar chart SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

**Comment by 1ambda** [comment]
rebased

**Comment by 1ambda** [comment]
@Leemoonsoo @felixcheung @bzz test is now green :)

**Comment by 1ambda** [comment]
@bzz Please let me know if you need anything to merge this PR :)

**Comment by bzz** [comment]
Looks great to me, 
merging to master if there is no further discussion

**Pull request closed without merging**